### ISSUE TITLE
💅 Biome reports false-positives on non-array's `.forEach` signatures

### ISSUE THREAD

**Issue by nikelborm** [biomejs/biome#10799]
### Environment information

<details>

```
CLI:
  Version:                      2.5.1
  Color support:                true

Platform:
  CPU Architecture:             x86_64
  OS:                           linux

Environment:
  BIOME_DISTRIBUTION:                npm
  BIOME_LOG_PATH:                    unset
  BIOME_LOG_PREFIX_NAME:             unset
  BIOME_LOG_LEVEL:                   unset
  BIOME_LOG_KIND:                    unset
  BIOME_CONFIG_PATH:                 unset
  BIOME_THREADS:                     unset
  BIOME_WATCHER_KIND:                unset
  BIOME_WATCHER_POLLING_INTERVAL:    unset
  NO_COLOR:                     unset
  TERM:                         xterm-256color
  JS_RUNTIME_VERSION:           v26.3.0
  JS_RUNTIME_NAME:              node
  NODE_PACKAGE_MANAGER:         bun/1.3.14

Biome Configuration:
  Status:                       Not set
  Path:                         unset

Workspace:
  Open Documents:               0
```

</details>

### Rule name

`useIterableCallbackReturn`

### Playground link

[https://biomejs.dev/playground/?lintRules=useIterableCallbackReturn&tab.....](https://biomejs.dev/playground/?lintRules=useIterableCallbackReturn&tab=formatter&pane=Diagnostics&code=aQBtAHAAbwByAHQAIAB7ACAAQwBvAG4AcwBvAGwAZQAsACAAQQByAHIAYQB5ACAAYQBzACAARQBBAHIAcgBhAHkALAAgAEUAZgBmAGUAYwB0ACwAIABmAGwAbwB3ACwAIABJAHQAZQByAGEAYgBsAGUALAAgAHAAaQBwAGUAIAB9ACAAZgByAG8AbQAgACIAZQBmAGYAZQBjAHQAIgA7AAoACgBhAHcAYQBpAHQAIABwAGkAcABlACgACgAJAEkAdABlAHIAYQBiAGwAZQAuAHIAYQBuAGcAZQAoADAALAAgADIAKQAsAAoACQBFAGYAZgBlAGMAdAAuAGYAbwByAEUAYQBjAGgAKAAKAAkACQAoAG4AdQBtACkAIAA9AD4ACgAJAAkACQBwAGkAcABlACgACgAJAAkACQAJAEUAZgBmAGUAYwB0AC4AcwB1AGMAYwBlAGUAZAAoAFMAdAByAGkAbgBnAC4AZgByAG8AbQBDAGgAYQByAEMAbwBkAGUAKAA2ADUAIAArACAAbgB1AG0AKQApACwACgAJAAkACQAJAEUAZgBmAGUAYwB0AC4AZABlAGwAYQB5ACgAYAAkAHsAbgB1AG0AfQAgAHMAZQBjAG8AbgBkAHMAYAApACwACgAJAAkACQAJAEUAZgBmAGUAYwB0AC4AdABhAHAAKABDAG8AbgBzAG8AbABlAC4AbABvAGcAKQAsAAoACQAJAAkAKQAsAAoACQAJAHsAIABjAG8AbgBjAHUAcgByAGUAbgBjAHkAOgAgACIAdQBuAGIAbwB1AG4AZABlAGQAIgAgAH0ALAAKAAkAKQAsAAoACQBFAGYAZgBlAGMAdAAuAGYAbABhAHQATQBhAHAAKABmAGwAbwB3ACgARQBBAHIAcgBhAHkALgBqAG8AaQBuACgAIgAiACkALAAgAEMAbwBuAHMAbwBsAGUALgBsAG8AZwApACkALAAKAAkARQBmAGYAZQBjAHQALgByAHUAbgBQAHIAbwBtAGkAcwBlACwACgApADsACgAKAC8ALwAgAHAAcgBpAG4AdABzADoACgAvAC8AIABBAAoALwAvACAAQgAKAC8ALwAgAEMACgAvAC8AIABBAEIAQwAKAAoALwAvACAAcAByAG8AbwBmACAAdABoAGEAdAAgAHQAaABpAHMAIABmAG8AcgBFAGEAYwBoACAAaQBtAHAAbABlAG0AZQBuAHQAYQB0AGkAbwBuACAAZQB4AHAAZQBjAHQAcwAgAHYAYQBsAHUAZQBzACAAdABvACAAYgBlACAAcgBlAHQAdQByAG4AZQBkACAAZgByAG8AbQAgAHQAaABlACAAYwBhAGwAbABiAGEAYwBrAA%3D%3D)

[repro repo](https://github.com/nikelborm/biome-foreach-false-positive-repro/tree/main)

### Expected result

Linter should look at the actual TypeScript signature to determine whether the callback is expected to return values or not.

The callback is expected to return a `void` value:
```ts
Array<never>.forEach(callbackfn: (value: never, index: number, array: never[]) => void, thisArg?: any): void
```
`useIterableCallbackReturn` is expected to flag returned values from the callback above.

The callback is expected to return a non-`void` value:
```ts
// from Effect.ts
declare const forEach: <B, E, R, S extends Iterable<any>>(
    f: (a: RA.ReadonlyArray.Infer<S>, i: number) => Effect<B, E, R>,
    options?: {
      readonly concurrency?: Concurrency | undefined
      readonly batching?: boolean | "inherit" | undefined
      readonly discard?: false | undefined
      readonly concurrentFinalizers?: boolean | undefined
    } | undefined
  ) => (
    self: S
  ) => Effect<RA.ReadonlyArray.With<S, B>, E, R>
```
No error should be present in this case.

### Additional info

The current issue is not a duplicate of these:

- https://github.com/biomejs/biome/issues/9445
- https://github.com/biomejs/biome/issues/8247

The current issue is about `useIterableCallbackReturn` being matched, it seems, exclusively based on the method name, like `.forEach`, and applied to non-arrays. The ones listed above stay applied in the correct array context.

### Code of Conduct

- [x] I agree to follow Biome's Code of Conduct

**Comment by ematipico** [comment]
We definitely won't be check the types to fix the rule, however we could add an option like we did for `noForEach`

https://github.com/biomejs/biome/pull/4498

**Comment by nikelborm** [comment]
Is it possible, instead of relying on types, to exclude `.forEach` calls on module objects ~~and plain non-instance-method functions, like just `forEach`~~?
Referencing one of the older issues, [#1205](https://github.com/biomejs/biome/issues/1205):

> The reasoning behind the rule was that if a forEach method is available, then the object should be iterable.

The cases provided below specifically address the scope where the module object ~~or a function declared at the root of the file~~ cannot possibly satisfy the iterable interface.

These can be determined without types, just with the AST, I think. Like excluding matches accessed like this:
1. `import { Effect } from "anylib"; Effect.forEach(...)`
2. `import * as Effect from "anylib/Effect"; Effect.forEach(...)`
3. ~~`import { forEach } from 'effect/Effect'; forEach(...)`~~
4. ~~`const forEach = ...; forEach(...)`~~

`Effect` at runtime in the 1st example is ESM's special `Module Namespace Exotic Object`, and during file parsing might be possible to match for such objects. (Might require working with ASTs of more than one file)

There's a reasonable chance that such imported from outside `forEach` ~~methods~~ functions will not conform "`forEach`'s callback returns `void`" convention


UPD: I checked that the rule doesn't trigger on plain `forEach` functions imported like `import { forEach } from 'effect/Effect'`, or defined like `const forEach = ...;`

UPD2: And the fact that plain `forEach` functions don't trigger the warning is even beneficial for my argument, because `Effect.forEach` is the same plain function, just passed through the import chain with an additional link.
